src/Components/Blazor/WebAssembly.Authentication/src/Services/IRemoteAuthenticationService.cs

```diff
@@ -0,0 +1,54 @@
+
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Authentication
+{
+    /// <summary>
+    /// Represents a contract for services that perform authentication operations for a Blazor webassembly application.
+    /// </summary>
+    /// <typeparam name="TRemoteAuthenticationState">The state to be persisted across authentication operations.</typeparam>
+    public interface IRemoteAuthenticationService<TRemoteAuthenticationState> : IAccessTokenProvider
+        where TRemoteAuthenticationState : RemoteAuthenticationState
+    {
+        /// <summary>
+        /// Signs in a user.
+        /// </summary>
+        /// <param name="context">The <see cref="RemoteAuthenticationContext{TRemoteAuthenticationState}"/> for authenticating the user.</param>
+        /// <returns>The result of the authentication operation.</returns>
+        Task<RemoteAuthenticationResult<TRemoteAuthenticationState>> SignInAsync(RemoteAuthenticationContext<TRemoteAuthenticationState> context);
+
+        /// <summary>
+        /// Completes the sign in operation for a user when it is performed outside of the application origin via a redirect operation followed
+        /// by a redirect callback to a page in the application.
+        /// </summary>
+        /// <param name="context">The <see cref="RemoteAuthenticationContext{TRemoteAuthenticationState}"/> for authenticating the user.</param>
+        /// <returns>The result of the authentication operation.</returns>
+        Task<RemoteAuthenticationResult<TRemoteAuthenticationState>> CompleteSignInAsync(
+            RemoteAuthenticationContext<TRemoteAuthenticationState> context);
+
+        /// <summary>
+        /// Signs out a user.
+        /// </summary>
+        /// <param name="context">The <see cref="RemoteAuthenticationContext{TRemoteAuthenticationState}"/> for authenticating the user.</param>
+        /// <returns>The result of the authentication operation.</returns>
+        Task<RemoteAuthenticationResult<TRemoteAuthenticationState>> SignOutAsync(
+            RemoteAuthenticationContext<TRemoteAuthenticationState> context);
+
+        /// <summary>
+        /// Completes the sign out operation for a user when it is performed outside of the application origin via a redirect operation followed
+        /// by a redirect callback to a page in the application.
+        /// </summary>
+        /// <param name="context">The <see cref="RemoteAuthenticationContext{TRemoteAuthenticationState}"/> for authenticating the user.</param>
+        /// <returns>The result of the authentication operation.</returns>
+        Task<RemoteAuthenticationResult<TRemoteAuthenticationState>> CompleteSignOutAsync(
+            RemoteAuthenticationContext<TRemoteAuthenticationState> context);
+
+        /// <summary>
+        /// Gets the currently authenticated user.
+        /// </summary>
+        Task<ClaimsPrincipal> GetCurrentUser();
```

Comment:
How inconvenient does it become if we didn't have this property? What if this is only exposed by `IAuthenticationStateProvider`?

I'm just concerned that we will have so many similarly-named and similar-purpose APIs that people will not know what to use, or whether they all hold the same info.

Reply:
done